Add `--fail-fast`

tgdwyer opened the pull request:
Added a command line option "--stop-on-fail".
When in --stop-on-fail mode, after a test in an example group fails, running tests from subsequent example groups is aborted.

sol commented on test/integration/failing-multiple-groups/Foo.hs:
```diff
@@ -0,0 +1,22 @@
+module Foo where
```
I think ideally you want to be able to guess from the path to what feature a specific test file relates to.  For that reason I think putting this under `test/ integration/fail-fast/`

sol commented:
Regarding the failing CI  job, you'll need to run `hpack` to update the `.cabal`-file.

sol commented:
@tgdwyer just to be sure, because the tests take a long time to run, you can  focus on individual test cases by changing `it` to `fit` (or alternatively focus on a group of test cases by changing `describe` to `fdescribe`.  This is usually more convenient then passing `--match` on the command line.

https://hackage.haskell.org/package/hspec-2.11.9/docs/Test-Hspec.html#g:5

tgdwyer commented on src/Options.hs:
```diff
@@ -28,14 +28,16 @@ import           Info
   , "  --fast         disable :reload between example groups"
   , "  --preserve-it  preserve the `it` variable between examples"
+  , "  --no-magic     disable magic mode"
+  , "  --stop-on-fail stop testing after the first failure"
```
done

tgdwyer commented:
I have renamed the option to --fail-fast and made the various merges and cleanups.
I haven't added the test case for errors in $setup just yet or investigated adding MaybeT to the Report type.  That seems a bit more involved for now but am interested to investigate when I have time in a couple of weeks, with changes to follow under a different PR.

Anything else?

sol commented:
> I did some experimenting with $setup and it looks like it fast fails already.

I think what's happening here is that, if set up fails, all examples from the current module will be skipped.  My suspicion is that it would still continue with other modules. So to reproduce the "issue" you would need a test case with multiple modules.

I somehow missed that on my first review.

tgdwyer commented:
> My suspicion is that it would still continue with other modules. So to reproduce the "issue" you would need a test case with multiple modules.
I've changed the logic of runModule to skip if failures > 0 (rather than only on increasing failure count).  Thus, after the first failure all subsequent examples will be skipped regardless of module.
I've added a second file to the test case to confirm it is skipped.

I've also run hpack so the test files are included in doctest.cabal and the CI should run.

sol commented on src/Options.hs:
```diff
@@ -28,14 +28,16 @@ import           Info
   , "  --fast         disable :reload between example groups"
   , "  --preserve-it  preserve the `it` variable between examples"
+  , "  --fail-fast    abort on first failure"
+  , "  --no-magic     disable magic mode"
```
👍

tgdwyer commented:
I think I've implemented all the requested changes and that it's ready for merge, if you're happy?

sol reviewed (changes requested):
I think this is still not working correctly if you have multiple `$setup` blocks (see my other comment).

Would you mind cherry-picking https://github.com/sol/doctest/commit/bbe00c656d9f598de5b71307bb9067cbda8f7351 and using this as a starting point?

sol commented:
@tgdwyer also cherry-pick https://github.com/sol/doctest/commit/b4a777c26f31b2ba8a686b4d1d494c0387384c83, or just rebase on https://github.com/sol/doctest/tree/maybe-t.

tgdwyer commented on src/Runner.hs:
```diff
@@ -169,6 +171,19 @@ runModule fastMode preserveIt repl (Module module_ setup examples) = do
+    runExamples (testGroup:moreGroups) = do
+      failures <- sFailures <$> getSummary
+      case failFast of
+        FailFast    -> when (failures == 0) runAndContinue
```
Yes!  In my latest commit I have an extra test for setup across modules

tgdwyer commented:
> @tgdwyer also cherry-pick [b4a777c](https://github.com/sol/doctest/commit/b4a777c26f31b2ba8a686b4d1d494c0387384c83), or just rebase on https://github.com/sol/doctest/tree/maybe-t.

Done! Thanks for wrapping Report in MaybeT!  FailFast logic is much better now.  Latest commits just use forM_ to traverse the Maybe effect in runModule, same as runModules.  Code is much cleaner.

sol reviewed (changes requested):
Great, the code looks good now 👍

Let's clean up the test descriptions, and then it's good to go.

In addition, I think you need to run hpack again.

tgdwyer commented on test/MainSpec.hs:
```diff
@@ -59,6 +64,22 @@ spec = do
+        (cases 3) {sTried = 3, sFailures = 1}
+
+    it "in --fail-fast mode, does not run subsequent groups after fail" $
+      doctestWithFailFast "." ["fail-fast/Foo.hs"]
```
my mistake!  Indeed, it wasn't running the imported tests properly.

tgdwyer commented:
Thank you @sol and Simon for your code reviews and suggestions.  This was
my first contribution to a major Haskell project - I learned a lot!

Kind regards,

Tim

On Fri, 18 Oct 2024 at 18:54, Simon Hengel ***@***.***> wrote:

> Merged #455 <https://github.com/sol/doctest/pull/455> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sol/doctest/pull/455#event-14728780620>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVAQ4KFE236WR27VNQTVP3Z4G3WPAVCNFSM6AAAAABPCKTG72VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJUG4ZDQNZYGA3DEMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Tim Dwyer BSc BCS(Hons) PhD
Professor of Data Visualisation and Immersive Analytics
Faculty of Information Technology
Monash University
Australia
http://marvl.infotech.monash.edu/~dwyer/
Mobile: 0481 240 767
(from overseas: +61 481 240 767)
